grails-app/controllers/org/pih/warehouse/api/FulfillmentApiController.groovy

```diff
@@ -0,0 +1,59 @@
+package org.pih.warehouse.api
+
+import grails.converters.JSON
+import grails.util.Holders
+import org.grails.plugins.web.taglib.ApplicationTagLib
+import org.pih.warehouse.CommandUtils
+import org.pih.warehouse.fulfillment.FulfillmentService
+import org.pih.warehouse.outbound.ImportPackingListCommand
+import org.pih.warehouse.outbound.ImportPackingListErrors
+import org.pih.warehouse.outbound.ImportPackingListItem
+
+class FulfillmentApiController {
+
+    FulfillmentService fulfillmentService
+
+    def validate(ImportPackingListCommand command) {
+        ImportPackingListErrors errors = new ImportPackingListErrors()
+        if (command.hasErrors()) {
+            response.status = 400
+            buildErrors(command, errors)
+        }
+        List<Map> tableData = command.packingList.collect { ImportPackingListItem item -> item.toTableJson() }
+        render([errors: errors, data: tableData] as JSON)
+    }
+
+    def createOutbound(ImportPackingListCommand command) {
+        ImportPackingListErrors errors = new ImportPackingListErrors()
+        if (command.hasErrors()) {
+            response.status = 400
+            buildErrors(command, errors)
+            List<Map> tableData = command.packingList.collect { ImportPackingListItem item -> item.toTableJson() }
+            render([errors: errors, data: tableData] as JSON)
+            return
+        }
+        StockMovement stockMovement = fulfillmentService.createOutbound(command)
+
+        render([data: stockMovement] as JSON)
+    }
+
+    static void buildErrors(ImportPackingListCommand command, ImportPackingListErrors errors) {
+        // Build errors instance for each sub-step (fulfillmentDetails, sendingOptions, list of items)
+        // Errors will be grouped by the sub-step and by fields, e.g.: { fulfillmentDetails: { origin: ["Cannot be null"] } }
+        errors.fulfillmentDetails = CommandUtils.buildErrorsGroupedByField(command.fulfillmentDetails.errors)
+        errors.sendingOptions = CommandUtils.buildErrorsGroupedByField(command.sendingOptions.errors)
+        // If the packing list is empty and validator has detected it, return generalErrors that will contain the info of packingList being empty
+        if (command.packingList.empty && command.errors.hasFieldErrors("packingList")) {
+            ApplicationTagLib g = Holders.grailsApplication.mainContext.getBean(ApplicationTagLib)
+            errors.packingList["generalErrors"] = command.errors.getFieldErrors("packingList").collect { g.message(error: it) }
+            return
+        }
+        command.packingList.each { ImportPackingListItem item ->
+            if (item.hasErrors()) {
+                // Build errors for each row separately. Row id is the key.
+                // e.g. { 2: { binLocation: ["Bin location cannot be null"] } }
```

Comment:
I could argue that having a map is not a standard way of rendering the data from an API. 
It is even one of the known patterns to normalize the API, like:
```
{
  allIds: [1,2,3],
  byId: {
    1: {...},
    2: {...},
    3: {...},
}
```
https://netflix.github.io/falcor/documentation/jsongraph.html

If you strongly disagree with that solution, we can think through that, and if it would be harmful to adjust frontend to eventual refactor.
Any other opinions on that? @alannadolny @awalkowiak @drodzewicz @EWaterman ?

Reply:
I think what Justin is getting at here is that we want a general solution to error handling that is agnostic of the frontend. The API should be able to return to the client saying "here are all your errors" and then let the client decide what to do with that. The fact that the frontend is rendering a table of data should be irrelevant to the backend. Another frontend could call this same API and render the result in a totally different way.

And if we can format our errors in a standardized way that will work for any of our APIs, then we can simplify a lot of our code going forward by building some quite smart general purpose error builder utils.

Taking Justin's idea a step further, we could do something like:
```
errors: [
    { 
        identifiers: {
            property: "binLocation",
            index/rowNumber: 1,
        },
        errorMessages: {
            default: "Bin location cannot be null",
            en: "Bin location cannot be null",
        }
    },
    { 
        identifiers: {
            property: "binLocation",
            index/rowNumber: 1,
        },
        errorMessages: {
            default: <Some other error on the same bin location field>,
            ...
        }
    },
]
```
Where "identifiers" is all the data required to identify the element. Structuring it flat like this allows us to more easily key the data by whatever we want since we can introduce new identifiers whenever needed without needing to add another layer of nested maps.

If we have another API that isn't returning data meant for rows, it could still use the same format:
```
errors: [
    { 
        identifiers: {
            property: "orderName",
            somethingElse: "blah",
        },
        errorMessages: {
            default: "Order name can't be empty",
            ...
        }
    },
    ...
]
```

All that said, I see your point @kchelstowski in that if we're returning errors for a table of data, it's very likely that we're going to be associating that data first and foremost with the row. If we ever want to look up errors just for a single row, then yes we'd need to loop through everything in this design, but I don't really know what the use case is for that scenario. I'd imagine we'd almost always want to loop through all errors so that we can show all errors for all rows at the same time. If we're only operating on a single row at once, we'd probably have a different API that only updates a single row, and thus only returns errors for that one row.

I think it's a valid concern, but I think it's outside of our use case currently, and the advantages that we'd get out of a more general purpose error format outweigh the drawbacks. As Justin suggested, in the worst case the frontend could massage the data into whatever format it needs.

But I'm still open to discussion on this.